pkg/auth/oauth2_client.go

```diff
@@ -0,0 +1,191 @@
+	}
+	if n > codeVerifierLenMax {
+		n = codeVerifierLenMax
+	}
+
+	// Randomly choose some allowed characters...
+	b := make([]byte, n)
+	for i := range b {
+		// ensure we use non-deterministic random ints.
+		j, _ := rand.Int(rand.Reader, big.NewInt(int64(len(codeVerifierAllowedLetters))))
+		b[i] = codeVerifierAllowedLetters[j.Int64()]
+	}
+
+	return string(b)
+}
+
+// generateCodeChallenge returns a standards compliant PKCE S(HA)256 code
+// challenge.
+func generateCodeChallenge(codeVerifier string) string {
+	// Create a sha-265 hash from the code verifier...
+	s256 := sha256.New()
+	s256.Write([]byte(codeVerifier))
+
+	// Then base64 encode the hash sum to create a code challenge...
+	return base64.RawURLEncoding.EncodeToString(s256.Sum(nil))
+}
+
+// resetPKCE cleans up PKCE details and returns the code verifier.
+func resetPKCE() (codeVerifier string) {
+	codeVerifier = PkceCodeVerifier
+	PkceCodeVerifier = ""
+	return codeVerifier
+}
+
+
+// A valid oauth2 client (check the store) that additionally requests an OpenID Connect id token
+var clientConf = goauth.Config{
+	ClientID:     "flytectl",
+	ClientSecret: "foobar",
+	RedirectURL:  "http://localhost:3846/callback",
+	Scopes:       []string{"all"},
+	Endpoint: goauth.Endpoint{
+		TokenURL: "http://localhost:8088/oauth2/token",
+		AuthURL:  "http://localhost:8088/oauth2/authorize",
+	},
+}
+
+type FlyteCtlTokenSource struct {
+	accessToken string
+}
+
+func (ts *FlyteCtlTokenSource) Token() (*goauth.Token, error) {
+	t := &goauth.Token{
+		AccessToken: ts.accessToken,
+		Expiry:      time.Now().Add(1 * time.Minute),
+		TokenType:   "bearer",
+	}
+	return t, nil
+}
+
+type InsecurePerRPCCredentials struct{
+	goauth.TokenSource
+}
+
+func (cr InsecurePerRPCCredentials) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+	token, err := cr.Token()
+	if err != nil {
+		return nil, err
+	}
+	return map[string]string{
+		"authorization": token.Type() + " " + token.AccessToken,
+	}, nil
+}
+
+func (cr InsecurePerRPCCredentials) RequireTransportSecurity() bool {
+	return false
+}
+
+func StartAuthFlow(ctx context.Context) (grpc.CallOption, error) {
+	// ### oauth2 client ###
+	http.HandleFunc("/callback", CallbackHandler(clientConf)) // the oauth2 callback endpoint
+	port := "3846"
+	TokenChannel = make(chan string)
+	if os.Getenv("FLYTE_AUTH_PORT") != "" {
+		port = os.Getenv("FLYTE_AUTH_PORT")
+	}
+	PkceCodeVerifier = generateCodeVerifier(64)
+	PkceCodeChallenge = generateCodeChallenge(PkceCodeVerifier)
+	urlToOpen := clientConf.AuthCodeURL("some-random-state-foobar") + "&nonce=some-random-nonce&code_challenge=" +
+		PkceCodeChallenge + "&code_challenge_method=S256"
+	go func() {
+		if err := http.ListenAndServe(":"+port, nil); err != nil {
+			logger.Fatal(ctx, "Couldn't start the callback http server on port %v", port)
+		}
+	}()
+	fmt.Println("Please open your webbrowser at " + urlToOpen)
+	_ = exec.Command("open", urlToOpen).Run()
```

Comment:
We should probably use this: https://pkg.go.dev/github.com/pkg/browser to ensure it works on all OSes

Reply:
Done